src/colorsampler.jl

```diff
@@ -0,0 +1,117 @@
+@enum Interpolation Linear Nearest
+
+struct Scaling{F, R}
+    # a function to scale a value by, e.g. log10, sqrt etc
+    scaling_function::F
+    # If nothing, no scaling applied!
+    range::R
+end
+
+Scaling() = Scaling(identity, nothing)
+
+const NoScaling = Scaling{typeof(identity), Nothing}
+
+struct Sampler{N, V} <: AbstractArray{RGBAf0, 1}
+    # the colors to sample from!
+    colors::AbstractArray{T, N} where T
+    # or an array of values, which are used to index into colors via interpolation!
+    values::V
+    # additional alpha that gets multiplied
+    alpha::Float64
+    # interpolation method for sampling
+    interpolation::Interpolation
+    scaling::Scaling
+end
+
+Base.size(sampler::Sampler) = size(sampler.values)
+
+"""
+    interpolated_getindex(cmap::AbstractArray, value::AbstractFloat, norm = (0.0, 1.0))
+
+Like getindex, but accepts values between 0..1 and interpolates those to the full range.
+You can use `norm`, to change the range of 0..1 to whatever you want.
+"""
+function interpolated_getindex(cmap::AbstractArray, value::Number, norm::VecTypes)
+    cmin, cmax = norm
+    i01 = clamp((value - cmin) / (cmax - cmin), 0.0, 1.0)
+    if !isfinite(i01)
+        i01 = 0.0
+    end
+    return interpolated_getindex(cmap, i01)
+end
+
+"""
+    interpolated_getindex(cmap::AbstractArray, value::AbstractFloat)
+
+Like getindex, but accepts values between 0..1 for `value` and interpolates those to the full range of `cmap`.
+"""
+function interpolated_getindex(cmap::AbstractArray{T}, i01::AbstractFloat) where T
+    i1len = (i01 * (length(cmap) - 1)) + 1
+    down = floor(Int, i1len)
+    up = ceil(Int, i1len)
+    down == up && return cmap[down]
+    interp_val = i1len - down
+    downc, upc = cmap[down], cmap[up]
+    return convert(T, (downc * (1.0 - interp_val)) + (upc * interp_val))
+end
+
+function nearest_getindex(cmap::AbstractArray, value::AbstractFloat)
+    idx = round(Int, i01 * (length(cmap) - 1)) + 1
+    return cmap[idx]
+end
+
+"""
+    apply_scaling(value::Number, scaling::Scaling)
+
+Scales a number to the range 0..1.
+"""
+function apply_scaling(value::Number, scaling::NoScaling)::Float64
+    return Float64(value)
+end
+
+function apply_scaling(value::Number, scaling::Scaling)::Float64
+    value_scaled = scaling.scaling_function(value)
```

Comment:
This would throw an error if given log(0) - would we need to handle this directly in `Sampler`?

Reply:
Good point!